File: sdk/src/sanitize.ts
/**
 * Sanitize context values to prevent prompt injection.
 * Escapes template tags and XML-like tags in string values.
 */
export function sanitizeContext(context: Record<string, any>): Record<string, any> {
  const sanitized: Record<string, any> = {}
  for (const [key, value] of Object.entries(context)) {
    if (typeof value === "string") {
      sanitized[key] = value
        .replace(/\{\{/g, "{ {")
        .replace(/\}\}/g, "} }")
    } else if (value && typeof value === "object" && !Array.isArray(value)) {
      sanitized[key] = sanitizeContext(value)
    } else {
      sanitized[key] = value
    }
  }
  return sanitized
}

File: sdk-go/sanitize.go
package composr

import "strings"

// sanitizeContext escapes template tags in string values to prevent prompt injection.
func sanitizeContext(ctx map[string]interface{}) map[string]interface{} {
	sanitized := make(map[string]interface{}, len(ctx))
	for k, v := range ctx {
		switch val := v.(type) {
		case string:
			s := strings.ReplaceAll(val, "{{", "{ {")
			s = strings.ReplaceAll(s, "}}", "} }")
			sanitized[k] = s
		case map[string]interface{}:
			sanitized[k] = sanitizeContext(val)
		default:
			sanitized[k] = v
		}
	}
	return sanitized
}
